pydantic/functional_validators.py

```diff
@@ -113,16 +114,25 @@ class Model(BaseModel):
     """
 
     func: core_schema.NoInfoValidatorFunction | core_schema.WithInfoValidatorFunction
+    input_type: Any = PydanticUndefined
 
-    def __get_pydantic_core_schema__(self, source_type: Any, handler: _GetCoreSchemaHandler) -> core_schema.CoreSchema:
+    def __get_pydantic_core_schema__(self, source_type: Any, handler: GetCoreSchemaHandler) -> core_schema.CoreSchema:
         schema = handler(source_type)
+        input_schema = None if self.input_type is PydanticUndefined else handler.generate_schema(self.input_type)
+        metadata = {'input_schema': input_schema}
```

Comment:
I'm using the `metadata` field to hold this information, instead of creating a new attribute in the relevant core schemas. That is because `input_schema` is only relevant for JSON Schema generation (unlike serializers where `return_schema` is actually used by `pydantic-core`).

Reply:
TODO I need to make proper use of the metadata field here (see `CoreMetadata` typed dict).

Reply:
Looking at `build_metadata_dict`, it looks like having extra metadata fields (that can be specified through `initial_metadata` argument is a supported pattern). So in fact I believe having this `input_schema` value added is fine?

Reply:
I think it would be better to not sprinkle a magic string `'input_schema'` around like this if at all possible, as it just makes it harder to keep track of all the different keys in the `metadata`. In contrast with adding a key to `build_metadata_dict` and using that function. But I don't feel _super_ strongly about this, if others (in particular, @sydney-runkle) prefer this approach

Reply:
I agree with @dmontagu, I would strongly prefer to avoid both of the above - I'd rather not add a new key, and I'd also like to avoid injecting magic strings into `metadata`. See my suggestion [here](https://github.com/pydantic/pydantic/pull/10094#issuecomment-2293653992) for a potential alternative approach.

Reply:
Ok, so following up on this - given that we are using the metadata to store this information, I'd rather that we sort of "officially" support it by adding this field to `CoreMetadata` and the `build_metadata_dict` constructor.

Reply:
It might be a bit slower / more verbose than actually needed, but I think it'll make a refactor down the line easier if we remain consistent with how we handle json schema metadata in this PR.